README.md
<div align="left">
    <img width="600" title="React Patterns" src="http://github.fork.de/ReactPatterns_201005.svg">
</div>

This repository contains basic React Components, Hooks and Patterns, which we use as a starting point at [Fork Unstable Media](https://www.fork.de/).

**Table of contents**

- [Features](#features)
- [Installation](#installation)
- [Theming](#theming)
- [Stylesheets](#stylesheets)
- [Deployment](#deployment)
- [Roadmap](#roadmap)

<!-- /TOC -->

---

## Features

- 📦 Basic React components to start with
- 💅 styled-components for styling
- 🔧 ESLint, Stylelint und Prettier already configured
- 🧱 Storybook Integration with A11y, Docs, Viewport and Theme Playground Plugin
- 📐 Styling helpers for color, spacings and media queries
- ⚙️ Basic jest snapshot testing is already implemented

## Installation

Clone the `react-patterns` repository.

```sh
git clone https://github.com/fork/react-patterns
cd react-patterns/
```

You can copy the components in two different ways to your React project.

### 1. Automatic migration

Run the `migrate.sh` script with your project path as an argument.

```sh
bash .migration/migrate.sh /Users/path/to/your/project
```

Follow the steps and the script will copy and install all necessary files and dependecies.

### 2. Manual migration

Copy all components you need for your new project. All components consist of a `component.jsx`, `component.style.jsx`, `component.stories.jsx` and a `index.js` file. All components depend on design tokens and our stylesheet helpers, therefore you also need to copy the `stylesheets` and the `tokens` directory to your new project.

<details>
  <summary>Install missing dependencies</summary>

```sh
# Install dependencies
yarn add react react-dom styled-components prop-types lazysizes picturefill what-input js-cookie @svgr/cli

# Install dev dependencies
yarn add -D @storybook/react @storybook/addon-a11y @storybook/theming @storybook/addon-docs @storybook/addon-viewport babel-loader @babel/core babel-jest enzyme enzyme-adapter-react-16 enzyme-to-json jest react-is react-test-renderer @4rk/staticpages-cli
```

</details>

<details>
  <summary>Add scripts to package.json</summary>

```sh
"storybook": "start-storybook -s ./public -p 8000",
"storybook:build": "build-storybook",
"deploy": "yarn storybook:build && staticpages-cli",
"test": "jest --config ./jest.config.json",
"test:update": "jest --config ./jest.config.json -u",
"build:icons": "svgr --no-dimensions --typescript -d src/components/Icon/compiled src/components/Icon/icons"
```

</details>

<details>
  <summary>Wrap your React tree in a <b>ThemeProvider</b></summary>

```js
import React from 'react';
import { ThemeProvider } from 'path/to/stylesheets';

export default () => (
  <ThemeProvider theme="default">
    <ReactTree />
  </ThemeProvider>
);
```

</details>

## Theming

The theming of our components is based on a `tokens` file, where we store our main variables. The `tokens` file consists of colors, spacings, breakpoints, grid sizes, font-sizes or animation properties. A basic example is located at `./tokens/index.js`.

You need to wrap the whole React tree inside a `ThemeProvider` to make the tokens available for the components. Otherwise the styling helpers will return an error.

```js
import { ThemeProvider } from 'path/to/stylesheets';

const App = ({ children }) => {
  return <ThemeProvider theme="default">{children}</ThemeProvider>;
};
```

## Stylesheets

We use `styled-components` to style our react components. There are several helper methods and tools in the `stylesheets` direcetory which will return theme aware design tokens.

**Example**

```js
import styled from 'styled-components';
import { color, space, mq } from 'path/to/stylesheets';

const StyledComponent = styled.h2`
  margin: ${space('m')} 0;
  color: ${color('primary')};

  ${mq('m')} {
    color: ${color('neutral', 30)};
  }
`;
```

## Deployment

1. Copy `.env.example` to `.env` and add your staticpages `DEPLOY_KEY`.
2. Run `yarn deploy` to deploy static storybook app via staticpages-cli.

## Roadmap

Write something about future releases and whats on the list for the next development phase.

- [x] Update readme
- [x] Use Typescript
- [ ] Add more jest tests

---

<div align="center">
  <img src="http://github.fork.de/heart.png" width="38" height="41" alt="Fork Logo" />

  <p>Brought to you by <a href="https://www.fork.de">Fork Unstable Media GmbH</a></p>
</div>

---


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

### Changed

### Removed

## [1.0.1] - 2021-02-01

### Added

- Added custom ThemeProvider
- Added tokens as CSS Variables
- Added TeaserContainer component
- Added Link component
- Basic Dialog component
- TypeScript linting
- Added Storybook Docs, enable to write MDX in separate file
- Added custom Storybook theme
- Added password component
- Added file upload component
- Added scroll container component
- Added IBM Plex Mono as default font
- Added Brand Guidelines dummy story

### Changed

- Refactored share links component
- Refactored Tabs to use buttons instead of anchors
- Updated babel config to build esmodules
- Added new Image Component
- Changed components to typescript
- Updated stories and jest snapshot tests for typescript
- Updated all packages
- Changed grid tokens structure & grid helpers
- Updated default template story
- Change Storybook hierarchy and order
- Refactored quote
- Refactored teaser

### Removed

- Removed svg-sprite-loader in favor of @svgr/cli
- Removed IconButton component in favor of iconPosition prop on Button
- Removed Cookielayer component
- Dedicated Picture component in favor of new Image Component
- Removed some unused packages (e.g. prop-types)
- Removed debug grid
- Removed components: box, flex, row and grid

## [0.1.0] - 2020-10-05


.migration/migrate.sh
#!/bin/bash

# Arguments
DIR=${@%/}
SRCDIR=${@%/src}

# Variables
DEP="react react-dom styled-components lazysizes picturefill what-input js-cookie @svgr/cli"
DEVDEP="@types/js-cookie @types/node @types/react @types/react-dom @types/styled-components typescript"

PURPLE='\033[0;35m'
BLUE='\033[0;34m'
GREEN='\033[0;32m'
NC='\033[0m'

INSTALL="${PURPLE}installing${NC}"
INFO="${BLUE}info${NC}"
SUCCESS="${GREEN}success${NC}"

JEST=false
STORYBOOK=false
STATICPAGES=false

if [ -z "$DIR" ]
then
  echo "You have to provide a directory as an argument,"
  echo "e.g. bash import.sh /users/path/to/directory"
  set -e
else  
  # Ask if user wants to copy the custom hooks directory
  while true; do
      read -p "Do you want to copy the 'hooks' directory? " yn
      case $yn in
          [Yy]* ) 
            cp -r -n ./src/hooks $SRCDIR; 
            break;;
          [Nn]* ) break;;
          
          * ) echo "Please insert yes or no";
      esac
  done

  # Ask if user wants to add storybook to their project
  while true; do
      read -p "Do you want to add Storybook to your directory? " yn
      case $yn in
          [Yy]* )
            cp -r -n ./.storybook $DIR; 
            STORYBOOK=true
            DEVDEP=$DEVDEP" @storybook/react @storybook/addon-a11y @storybook/theming @storybook/addon-docs @storybook/addon-viewport storybook-addon-theme-playground babel-loader @babel/core"; 
            break;;
          
          [Nn]* ) break;;
          
          * ) echo "Please insert yes or no";
      esac
  done

  # Ask if user wants to copy and add jest testing to new project
  while true; do
      read -p "Do you want to add Jest tests to your directory? " yn
      case $yn in
          [Yy]* ) 
            cp -r -n ./__tests__ $DIR; 
            cp -r -n ./setup $DIR; 
            cp ./jest.config.json $DIR; 
            
            JEST=true
            DEVDEP=$DEVDEP"@types/jest babel-jest enzyme enzyme-adapter-react-16 enzyme-to-json jest react-is react-test-renderer @babel/preset-typescript";
            
            break;;
          
          [Nn]* ) break;;
          
          * ) echo "Please insert yes or no";
      esac
  done

  # Ask if user wants to add staticpages-cli to their project
  while true; do
      read -p "Do you want to add the staticpages-cli for deployment? " yn
      case $yn in
          [Yy]* ) 
            STATICPAGES=true
            DEVDEP=$DEVDEP" @4rk/staticpages-cli";
            break;;
          
          [Nn]* ) break;;
          
          * ) echo "Please insert yes or no";
      esac
  done

  echo ""
  echo -e "$INFO Copying 'components', 'templates', 'stylesheets' directories to $SRCDIR."
  cp -r -n ./src/components $SRCDIR
  cp -r -n ./src/templates $SRCDIR
  cp -r -n ./src/stylesheets $SRCDIR
  cp -r -n ./src/tokens $SRCDIR
  echo -e "$INFO Copying public' directory to $DIR."
  cp -r -n ./public $DIR


  # Check if package.json exists, else init project
  PACKAGE="$DIR/package.json"
  if [ -f "$PACKAGE" ]; then
    echo ""
    echo -e "$INSTALL package.json found"
  else
    echo ""
    echo -e "$INSTALL Init project:"
    yarn --cwd $DIR init
  fi

  # Install dependencies
  echo ""
  echo -e "$INSTALL dependencies:"
  echo $DEP
  echo ""
  yarn --cwd $DIR add $DEP

  # Check if dev dependencies need to be installed
  if [ -z "$DEVDEP" ]
  then
    echo ""
    echo -e "$INSTALL No dev dependencies to install."
  else
    echo ""
    echo -e "$INSTALL dev dependencies:"
    echo $DEVDEP
    echo ""
    yarn --cwd $DIR add -D $DEVDEP
  fi

  if [ "$STORYBOOK" = true ] || [ "$JEST" = true ] || [ "$STATICPAGES" = true ] ; then
    echo ""
    echo -e "$INFO Add the follwing scripts to your package.json file:"; 
    echo ""

    if [ "$STORYBOOK" = true ] ; then
      echo "\"storybook\": \"start-storybook -s ./public -p 8000\","; 
      echo "\"storybook:build\": \"build-storybook\","; 
    fi

    if [ "$STATICPAGES" = true ] ; then
      echo "\"deploy\": \"yarn storybook:build && staticpages-cli\","; 
    fi

      
    if [ "$JEST" = true ] ; then
      echo "\"test\": \"jest --config ./jest.config.json\","; 
      echo "\"test:update\": \"jest --config ./jest.config.json -u\""; 
    fi
  fi


  if [ "$STATICPAGES" = true ] ; then
    echo ""
    echo -e "$INFO Add the follwing variables to your .env file:"; 
    echo ""
    echo "DEPLOY_KEY=staticpages-deploy-key";
    echo "DEPLOY_SRC=storybook-static";
  fi

  # Theme Provider Information
  echo ""
  echo -e "$INFO Also make sure that you wrap your React tree in a 'ThemeProvider':"; 
  echo ""
  echo "import React from 'react';"
  echo ""
  echo "import { ThemeProvider } from 'path/to/stylesheets';"
  echo ""
  echo "export default () => ("
  echo "  <ThemeProvider>"
  echo "    <ReactTree />"
  echo "  </ThemeProvider>"
  echo ");"

  echo ""
  echo -e "$SUCCESS Everything was copied and installed!"
  echo ""
fi